src/common/column_matrix.h

```diff
@@ -25,36 +26,42 @@ enum ColumnType {
 
 /*! \brief a column storage, to be used with ApplySplit. Note that each
     bin id is stored as index[i] + index_base. */
+template <typename T>
 class Column {
  public:
-  Column(ColumnType type, const uint32_t* index, uint32_t index_base,
-         const size_t* row_ind, size_t len)
+  Column(ColumnType type, const T* index, uint32_t index_base,
+         const size_t* row_ind, size_t len,
+         const std::vector<bool>* missing_flags, const size_t disp)
       : type_(type),
         index_(index),
         index_base_(index_base),
         row_ind_(row_ind),
-        len_(len) {}
+        len_(len),
+        missing_flags_(missing_flags), disp_(disp) {}
   size_t Size() const { return len_; }
-  uint32_t GetGlobalBinIdx(size_t idx) const { return index_base_ + index_[idx]; }
-  uint32_t GetFeatureBinIdx(size_t idx) const { return index_[idx]; }
-  common::Span<const uint32_t> GetFeatureBinIdxPtr() const { return { index_, len_ }; }
+  uint32_t GetGlobalBinIdx(size_t idx) const { return index_base_ + (uint32_t)(index_[idx]); }
+  T GetFeatureBinIdx(size_t idx) const { return index_[idx]; }
+  common::Span<const T> GetFeatureBinIdxPtr() const { return { index_, len_ }; }
   // column.GetFeatureBinIdx(idx) + column.GetBaseIdx(idx) ==
   // column.GetGlobalBinIdx(idx)
   uint32_t GetBaseIdx() const { return index_base_; }
   ColumnType GetType() const { return type_; }
   size_t GetRowIdx(size_t idx) const {
     // clang-tidy worries that row_ind_ might be a nullptr, which is possible,
     // but low level structure is not safe anyway.
-    return type_ == ColumnType::kDenseColumn ? idx : row_ind_[idx];  // NOLINT
+    return type_ == ColumnType::kDenseColumn ? idx : row_ind_[idx];// NOLINT
   }
-  bool IsMissing(size_t idx) const {
-    return index_[idx] == std::numeric_limits<uint32_t>::max();
+  inline bool IsMissing(size_t idx) const {
+    return (*missing_flags_)[disp_ + idx] == true;
   }
   const size_t* GetRowData() const { return row_ind_; }
 
+  const std::vector<bool>* missing_flags_;
+  const size_t disp_;
+
  private:
   ColumnType type_;
-  const uint32_t* index_;
+  const T* index_;
```

Comment:
Can we use a `Span` to replace these pointers and `index_base_`?

Reply:
Please correct me if I'm mistaken but as I understand `Span` contains row pointer and its size, but the meaning of  `index_base_`  is offset in histogram matrix for specific feature. If we add `_index_base` to all elements in `index_` we will lose opportunity to reduce memory consumption for `Column` . I think this fields are reasonably separated.  And we already have `GetFeatureBinIdxPtr()`  method which return `Span` instance.